src/Integration/LocalServices/UnboundProjectFinder.cs

```diff
@@ -71,34 +84,81 @@ public IEnumerable<Project> GetUnboundProjects()
 
         private Project[] GetUnboundProjects(BindingConfiguration binding)
         {
+            threadHandling.ThrowIfOnUIThread();
+
+            Debug.Assert(binding.Mode.IsInAConnectedMode());
+            Debug.Assert(binding.Project != null);
+
+            // Using threadHandling.Run(...) here to avoid having to convert all of the upstream
+            // calling methods to be async.
+            var projects = threadHandling.Run(() => GetUnboundProjectsAsync(binding));
```

Comment:
Side note: perhaps for readability we should rename `Run` to something more explicit, i.e. `RunOnCurrentThread`, to go along with `RunOnUIThread`